// @pexip/components/design-tokens/build-scripts/breakpoints.mjs
import fs from 'fs';
import path from 'path';

import {Content} from './content.mjs';

const generatedByFile =
    '/* FILE GENERATED BY packages/components/design-tokens/build-scripts/breakpoints.js DO NOT CHANGE */';

const breakpoints = {
    xs: 0,
    sm: 479,
    md: 744,
    lg: 1025,
    xl: 2000,
};

export function getBreakpoints() {
    return breakpoints;
}

function exportBreakpoints(breakpoints) {
    const content = new Content(generatedByFile);
    content.push('export const breakpoints = ');
    content.push(JSON.stringify(breakpoints));
    content.push(';');

    content.push('export type Breakpoint = keyof typeof breakpoints;');

    content.push('export const breakpointNames: Breakpoint[] = ');
    content.push(JSON.stringify(Object.keys(breakpoints)));
    content.push(';');

    fs.writeFileSync(
        path.resolve('..', 'generated', 'breakpoints.ts'),
        content.print('typescript'),
    );
}

export function syncBreakpoints(breakpoints) {
    const content = new Content(generatedByFile);
    content.push('$grid-breakpoints: (');
    for (const name in breakpoints) {
        const value = breakpoints[name];
        content.push(`'${name}': ${value}${value !== 0 ? 'px' : ''},`);
    }
    content.push(') !default;');
    fs.writeFileSync(
        path.resolve('..', 'generated', 'breakpoints.variables.scss'),
        content.print('scss'),
    );

    exportBreakpoints(breakpoints);
}
