// @pexip/media-processor/dist/main/benchUtils.js
export const createBenchmark = (clock = performance, { calculationThresholdMS = 1000 } = {}) => {
    const props = {
        beginTime: 0,
        endTime: 0,
        sumDelta: 0,
        count: 0,
        lastCalculateTime: 0,
        lastResult: 0,
    };
    return {
        begin: () => {
            props.beginTime = clock.now();
        },
        end: () => {
            props.endTime = clock.now();
            props.sumDelta += props.endTime - props.beginTime;
            ++props.count;
        },
        calculateFps: () => {
            const calculateTime = clock.now();
            const lastDuration = calculateTime - props.lastCalculateTime;
            if (lastDuration >= calculationThresholdMS) {
                const result = props.count === 0
                    ? 0
                    : (lastDuration * props.count) / props.sumDelta;
                props.sumDelta = 0;
                props.count = 0;
                props.lastCalculateTime = calculateTime;
                props.lastResult = result;
                return [calculateTime, result];
            }
            return [props.lastCalculateTime, props.lastResult];
        },
    };
};
