// @pexip/peer-connection-stats/dist/tests/statsCollector.firefox.fixtures.js
export const statsData = [
    {
        id: '70df0599',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: '32ac0d0c',
        nominated: false,
        priority: 8774140172997362000,
        readable: true,
        remoteCandidateId: '6479453f',
        selected: false,
        state: 'inprogress',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: 'cb3ebd3b',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: 'eab7c196',
        nominated: false,
        priority: 8774140172997231000,
        readable: true,
        remoteCandidateId: '6479453f',
        selected: false,
        state: 'inprogress',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: '747cde3e',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 5555374,
        bytesSent: 12060851,
        lastPacketReceivedTimestamp: 1607468216444,
        lastPacketSentTimestamp: 1607468216464,
        localCandidateId: '8f940ef2',
        nominated: true,
        priority: 7962083765516764000,
        readable: true,
        remoteCandidateId: '6479453f',
        selected: true,
        state: 'succeeded',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: '53941a10',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: '8f940ef2',
        nominated: true,
        priority: 7961802290540054000,
        readable: true,
        remoteCandidateId: '6479453f',
        selected: false,
        state: 'succeeded',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: 'f95809af',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: '85c269b',
        nominated: false,
        priority: 3676066494292132400,
        readable: true,
        remoteCandidateId: '62b88a85',
        selected: false,
        state: 'cancelled',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: 'fbd8e264',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: 'c715fb9b',
        nominated: false,
        priority: 3676066494292001300,
        readable: true,
        remoteCandidateId: '62b88a85',
        selected: false,
        state: 'cancelled',
        transportId: 'transport_0',
        writable: true,
    },
    {
        id: '92d8e29c',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: 'efec9d54',
        nominated: false,
        priority: 8774140172997362000,
        readable: true,
        remoteCandidateId: '4af8c58e',
        selected: false,
        state: 'inprogress',
        transportId: 'transport_1',
        writable: true,
    },
    {
        id: 'd4d8f4af',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: '93fefccb',
        nominated: false,
        priority: 8774140172997231000,
        readable: true,
        remoteCandidateId: '4af8c58e',
        selected: false,
        state: 'cancelled',
        transportId: 'transport_1',
        writable: true,
    },
    {
        id: '367faebf',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 415969,
        bytesSent: 163308,
        lastPacketReceivedTimestamp: 1607468216459,
        lastPacketSentTimestamp: 1607468216456,
        localCandidateId: '8f940ef2',
        nominated: true,
        priority: 7962083765516764000,
        readable: true,
        remoteCandidateId: '4af8c58e',
        selected: true,
        state: 'succeeded',
        transportId: 'transport_1',
        writable: true,
    },
    {
        id: 'd43ca551',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: 'ec5c247d',
        nominated: false,
        priority: 3676066494292132400,
        readable: true,
        remoteCandidateId: '53c089e',
        selected: false,
        state: 'cancelled',
        transportId: 'transport_1',
        writable: true,
    },
    {
        id: '5f7cb429',
        timestamp: 1607359512685,
        type: 'candidate-pair',
        bytesReceived: 0,
        bytesSent: 0,
        lastPacketReceivedTimestamp: 0,
        lastPacketSentTimestamp: 0,
        localCandidateId: '465c49cb',
        nominated: false,
        priority: 3676066494292001300,
        readable: true,
        remoteCandidateId: '53c089e',
        selected: false,
        state: 'cancelled',
        transportId: 'transport_1',
        writable: true,
    },
    {
        id: 'eab7c196',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.1.23',
        candidateType: 'host',
        port: 40524,
        priority: 2122187007,
        protocol: 'udp',
    },
    {
        id: '32ac0d0c',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.122.1',
        candidateType: 'host',
        port: 35588,
        priority: 2122252543,
        protocol: 'udp',
    },
    {
        id: 'c715fb9b',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.1.23',
        candidateType: 'host',
        port: 51647,
        priority: 2105458943,
        protocol: 'tcp',
    },
    {
        id: '85c269b',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.122.1',
        candidateType: 'host',
        port: 49524,
        priority: 2105524479,
        protocol: 'tcp',
    },
    {
        id: '8f940ef2',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '45.152.51.102',
        candidateType: 'prflx',
        port: 55186,
        priority: 1853817087,
        protocol: 'udp',
    },
    {
        id: '6479453f',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40166,
        priority: 2042888703,
        protocol: 'udp',
    },
    {
        id: '62b88a85',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40166,
        priority: 855900927,
        protocol: 'tcp',
    },
    {
        id: 'fbec0791',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40166,
        priority: 864289791,
        protocol: 'tcp',
    },
    {
        id: '93fefccb',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.1.23',
        candidateType: 'host',
        port: 37715,
        priority: 2122187007,
        protocol: 'udp',
    },
    {
        id: 'efec9d54',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.122.1',
        candidateType: 'host',
        port: 55186,
        priority: 2122252543,
        protocol: 'udp',
    },
    {
        id: '465c49cb',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.1.23',
        candidateType: 'host',
        port: 53452,
        priority: 2105458943,
        protocol: 'tcp',
    },
    {
        id: 'ec5c247d',
        timestamp: 1607359512685,
        type: 'local-candidate',
        address: '192.168.122.1',
        candidateType: 'host',
        port: 60325,
        priority: 2105524479,
        protocol: 'tcp',
    },
    {
        id: '4af8c58e',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40186,
        priority: 2042888703,
        protocol: 'udp',
    },
    {
        id: '53c089e',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40186,
        priority: 855900927,
        protocol: 'tcp',
    },
    {
        id: '58f0027f',
        timestamp: 1607359512685,
        type: 'remote-candidate',
        address: '34.91.46.168',
        candidateType: 'host',
        port: 40186,
        priority: 864289791,
        protocol: 'tcp',
    },
    {
        id: '8adcb9ce',
        timestamp: 1607359512687,
        type: 'inbound-rtp',
        kind: 'video',
        mediaType: 'video',
        ssrc: 840745016,
        discardedPackets: 0,
        jitter: 0.014,
        packetsLost: 0,
        packetsReceived: 4666,
        bitrateMean: 789199.2641509434,
        bitrateStdDev: 172342.0917917606,
        bytesReceived: 5215634,
        firCount: 0,
        framerateMean: 24.03773584905661,
        framerateStdDev: 4.823668665074425,
        framesDecoded: 1263,
        nackCount: 0,
        pliCount: 1,
        remoteId: '7b317e3f',
    },
    {
        id: 'b7b8198b',
        timestamp: 1607359512687,
        type: 'inbound-rtp',
        kind: 'audio',
        mediaType: 'audio',
        ssrc: 3376419831,
        jitter: 0,
        packetsLost: 0,
        packetsReceived: 2563,
        bytesReceived: 371213,
        nackCount: 0,
        remoteId: 'e1a8cb41',
    },
    {
        id: '9b81861',
        timestamp: 1607359512686,
        type: 'outbound-rtp',
        kind: 'video',
        mediaType: 'video',
        ssrc: 678033974,
        bytesSent: 11840687,
        packetsSent: 10663,
        bitrateMean: 1730236.3773584904,
        bitrateStdDev: 434797.7326515535,
        droppedFrames: 4,
        firCount: 0,
        framerateMean: 28.999999999999993,
        framerateStdDev: 5.844129204068486,
        framesEncoded: 1512,
        nackCount: 0,
        pliCount: 0,
        qpSum: 89087,
        remoteId: 'b39ceeda',
    },
    {
        id: '9b74fd1c',
        timestamp: 1607359512687,
        type: 'outbound-rtp',
        kind: 'audio',
        mediaType: 'audio',
        ssrc: 3453750983,
        bytesSent: 120316,
        packetsSent: 2560,
        nackCount: 0,
        remoteId: '8471050b',
    },
    {
        id: 'b39ceeda',
        timestamp: 1607359512666,
        type: 'remote-inbound-rtp',
        kind: 'video',
        mediaType: 'video',
        ssrc: 678033974,
        jitter: 0.002,
        packetsLost: 0,
        packetsReceived: 10497,
        bytesReceived: 11445691,
        localId: '9b81861',
        roundTripTime: 0.033,
    },
    {
        id: '8471050b',
        timestamp: 1607359510713,
        type: 'remote-inbound-rtp',
        kind: 'audio',
        mediaType: 'audio',
        ssrc: 3453750983,
        jitter: 0,
        packetsLost: 0,
        packetsReceived: 2232,
        bytesReceived: 78116,
        localId: '9b74fd1c',
        roundTripTime: 0.033,
    },
    {
        id: '7b317e3f',
        timestamp: 1607359512666,
        type: 'remote-outbound-rtp',
        kind: 'video',
        mediaType: 'video',
        ssrc: 840745016,
        bytesSent: 5305963,
        packetsSent: 4805,
        localId: '8adcb9ce',
        remoteTimestamp: 3816457015643,
    },
    {
        id: 'e1a8cb41',
        timestamp: 1607359510713,
        type: 'remote-outbound-rtp',
        kind: 'audio',
        mediaType: 'audio',
        ssrc: 3376419831,
        bytesSent: 347333,
        packetsSent: 2617,
        localId: 'b7b8198b',
        remoteTimestamp: 3816457014478,
    },
];
